### Importing Library:

In [1]:
import pandas as pd

### Loading "Matches" Dataset:

In [5]:
data1 = pd.read_csv("matches.csv")
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
data2 = pd.read_csv('deliveries.csv')
data1.head()

id  season       city        date                        team1  \
0   1    2017  Hyderabad  2017-04-05          Sunrisers Hyderabad   
1   2    2017       Pune  2017-04-06               Mumbai Indians   
2   3    2017     Rajkot  2017-04-07                Gujarat Lions   
3   4    2017     Indore  2017-04-08       Rising Pune Supergiant   
4   5    2017  Bangalore  2017-04-08  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN

### Filtering RR and MI matches only:

In [6]:
rr_mi = data1.loc[((data1['team1'] == 'Mumbai Indians') & (data1['team2'] == 'Rajasthan Royals') | (data1['team1'] == 'Rajasthan Royals') & (data1['team2'] == 'Mumbai Indians'))]
rr_mi.head()

id  season    city        date             team1             team2  \
85    86    2008  Mumbai  2008-05-07  Rajasthan Royals    Mumbai Indians   
112  113    2008  Jaipur  2008-05-26    Mumbai Indians  Rajasthan Royals   
159  160    2009  Durban  2009-05-14  Rajasthan Royals    Mumbai Indians   
175  176    2010  Mumbai  2010-03-13    Mumbai Indians  Rajasthan Royals   
218  219    2010  Jaipur  2010-04-11    Mumbai Indians  Rajasthan Royals   

          toss_winner toss_decision  result  dl_applied            winner  \
85     Mumbai Indians         field  normal           0    Mumbai Indians   
112  Rajasthan Royals         field  normal           0  Rajasthan Royals   
159  Rajasthan Royals           bat  normal           0  Rajasthan Royals   
175    Mumbai Indians           bat  normal           0    Mumbai Indians   
218  Rajasthan Royals         field  normal           0    Mumbai Indians   

     win_by_runs  win_by_wickets player_of_match                       venue  \
85             0               7         A Nehra  Dr DY Patil Sports Academy   
112            0               5   Sohail Tanvir      Sawai Mansingh Stadium   
159            2               0        SK Warne                   Kingsmead   
175            4               0       YK Pathan           Brabourne Stadium   
218           37               0    SR Tendulkar      Sawai Mansingh Stadium   

         umpire1      umpire2 umpire3  
85     DJ Harper  RE Koertzen     NaN  
112    BF Bowden  K Hariharan     NaN  
159  BR Doctrove    DJ Harper     NaN  
175  RE Koertzen    RB Tiffin     NaN  
218  BR Doctrove  SK Tarapore     NaN

### Question 1: Who will win the match?

#### Solution: Checking counts of winner:

In [9]:
rr_mi['winner'].value_counts()/rr_mi.shape[0] * 100

Rajasthan Royals    50.0
Mumbai Indians      50.0
Name: winner, dtype: float64

**Answer:** As per the above counts, RR and MI both won 10-10 matches.

## Question 2: How balls will Rashid Khan tak to get his first wicket?

### Loading "Deliveries" dataset:

In [9]:
data2 = pd.read_csv("deliveries.csv")
data2.head()

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
2         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
3         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
4         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  wide_runs  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0          0         0   
1     2  DA Warner    S Dhawan  TS Mills              0          0         0   
2     3  DA Warner    S Dhawan  TS Mills              0          0         0   
3     4  DA Warner    S Dhawan  TS Mills              0          0         0   
4     5  DA Warner    S Dhawan  TS Mills              0          2         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   
2            0            0             0             4           0   
3            0            0             0             0           0   
4            0            0             0             0           2   

   total_runs player_dismissed dismissal_kind fielder  
0           0              NaN            NaN     NaN  
1           0              NaN            NaN     NaN  
2           4              NaN            NaN     NaN  
3           0              NaN            NaN     NaN  
4           2              NaN            NaN     NaN

#### Solution: Filtering data with Rashid Khan as bowler:

In [22]:
rashid_wickets = data2.loc[(data2['bowler'] == "Rashid Khan")]
rashid_wicket

match_id  inning                 batting_team         bowling_team  \
157            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
158            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
159            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
160            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
161            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
162            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
169            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
170            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
171            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
172            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
173            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
174            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
202            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
203            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
204            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
205            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
206            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
207            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
214            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
215            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
216            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
217            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
218            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
219            1       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
1232           6       1                Gujarat Lions  Sunrisers Hyderabad   
1233           6       1                Gujarat Lions  Sunrisers Hyderabad   
1234           6       1                Gujarat Lions  Sunrisers Hyderabad   
1235           6       1                Gujarat Lions  Sunrisers Hyderabad   
1236           6       1                Gujarat Lions  Sunrisers Hyderabad   
1237           6       1                Gujarat Lions  Sunrisers Hyderabad   
1244           6       1                Gujarat Lions  Sunrisers Hyderabad   
1245           6       1                Gujarat Lions  Sunrisers Hyderabad   
1246           6       1                Gujarat Lions  Sunrisers Hyderabad   
1247           6       1                Gujarat Lions  Sunrisers Hyderabad   
1248           6       1                Gujarat Lions  Sunrisers Hyderabad   
1249           6       1                Gujarat Lions  Sunrisers Hyderabad   
1256           6       1                Gujarat Lions  Sunrisers Hyderabad   
1257           6       1                Gujarat Lions  Sunrisers Hyderabad   
1258           6       1                Gujarat Lions  Sunrisers Hyderabad   
1259           6       1                Gujarat Lions  Sunrisers Hyderabad   
1260           6       1                Gujarat Lions  Sunrisers Hyderabad   
1261           6       1                Gujarat Lions  Sunrisers Hyderabad   
1274           6       1                Gujarat Lions  Sunrisers Hyderabad   
1275           6       1                Gujarat Lions  Sunrisers Hyderabad   
1276           6       1                Gujarat Lions  Sunrisers Hyderabad   
1277           6       1                Gujarat Lions  Sunrisers Hyderabad   
1278           6       1                Gujarat Lions  Sunrisers Hyderabad   
1279           6       1                Gujarat Lions  Sunrisers Hyderabad   
2266          10       2               Mumbai Indians  Sunrisers Hyderabad   
2267          10       2               Mumbai Indians  Sunrisers Hyderabad   
2268          10       2      

#### Getting mean of the wide runs match wise:

In [12]:
wide = pd.DataFrame(srh_csk_wide.groupby('match_id')['wide_runs'].sum())
wide['wide_runs'].mean()

7.7727272727272725

**Answer:** Average wide runs in each match is 7.77. So, I would prdict that around 8 wides will be bowled in this match.

### Filtering SRH and CSK matches where batting team is SRH and bowling team is CSK:

In [19]:
srh_csk_bowl = data2.loc[(data2['batting_team'] == "Sunrisers Hyderabad") & (data2['bowling_team'] == "Chennai Super Kings")]
srh_csk_bowl.head(20)

match_id  inning         batting_team         bowling_team  over  ball  \
97960       414       1  Sunrisers Hyderabad  Chennai Super Kings     1     1   
97961       414       1  Sunrisers Hyderabad  Chennai Super Kings     1     2   
97962       414       1  Sunrisers Hyderabad  Chennai Super Kings     1     3   
97963       414       1  Sunrisers Hyderabad  Chennai Super Kings     1     4   
97964       414       1  Sunrisers Hyderabad  Chennai Super Kings     1     5   
97965       414       1  Sunrisers Hyderabad  Chennai Super Kings     1     6   
97966       414       1  Sunrisers Hyderabad  Chennai Super Kings     2     1   
97967       414       1  Sunrisers Hyderabad  Chennai Super Kings     2     2   
97968       414       1  Sunrisers Hyderabad  Chennai Super Kings     2     3   
97969       414       1  Sunrisers Hyderabad  Chennai Super Kings     2     4   
97970       414       1  Sunrisers Hyderabad  Chennai Super Kings     2     5   
97971       414       1  Sunrisers Hyderabad  Chennai Super Kings     2     6   
97972       414       1  Sunrisers Hyderabad  Chennai Super Kings     2     7   
97973       414       1  Sunrisers Hyderabad  Chennai Super Kings     2     8   
97974       414       1  Sunrisers Hyderabad  Chennai Super Kings     3     1   
97975       414       1  Sunrisers Hyderabad  Chennai Super Kings     3     2   
97976       414       1  Sunrisers Hyderabad  Chennai Super Kings     3     3   
97977       414       1  Sunrisers Hyderabad  Chennai Super Kings     3     4   
97978       414       1  Sunrisers Hyderabad  Chennai Super Kings     3     5   
97979       414       1  Sunrisers Hyderabad  Chennai Super Kings     3     6   

         batsman non_striker     bowler  is_super_over  wide_runs  bye_runs  \
97960  Q de Kock    S Dhawan  MM Sharma              0          0         0   
97961  Q de Kock    S Dhawan  MM Sharma              0          0         0   
97962  Q de Kock    S Dhawan  MM Sharma              0          0         0   
97963  GH Vihari    S Dhawan  MM Sharma              0          0         0   
97964  GH Vihari    S Dhawan  MM Sharma              0          0         0   
97965   S Dhawan   GH Vihari  MM Sharma              0          0         0   
97966   S Dhawan   GH Vihari  JO Holder              0          1         0   
97967   S Dhawan   GH Vihari  JO Holder              0          0         0   
97968   S Dhawan   GH Vihari  JO Holder              0          1         0   
97969   S Dhawan   GH Vihari  JO Holder              0          0         0   
97970  GH Vihari    S Dhawan  JO Holder              0          0         0   
97971   S Dhawan   GH Vihari  JO Holder              0          0         0   
97972   S Dhawan   GH Vihari  JO Holder              0          0         0   
97973   S Dhawan   GH Vihari  JO Holder              0          0         0   
97974   S Dhawan    CL White  MM Sharma              0          0         0   
97975   S Dhawan    CL White  MM Sharma              0          0         0   
97976   S Dhawan    CL White  MM Sharma              0          0         0   
97977   S Dhawan    CL White  MM Sharma              0          0         0   
97978   S Dhawan    CL White  MM Sharma              0          0         0   
97979   CL White    S Dhawan  MM Sharma              0          0         0   

       legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
97960            0            0             0             0           0   
97961            0            0             0             4           0   
97962            0            0             0             0           0   
97963            0            0             0             0           0   
97964            0            0             0             1           0   
97965            0            0             0             1           0   
97966            0            0             0             0           1   
97967            0            0             0

## Question 3: Which individual batsman in the match will have the best (number of runs scored of boundaries : (total runs scored) ratio?

#### Solution: Observing match wise David Warner's dismissal kind:

In [18]:
david_dismissal = pd.DataFrame(data2[data2['batsman'] == 'DA Warner'].groupby('dismissal_kind')['player_dismissed'].count())
david_dismissal = david_dismissal.rename({'player_dismissed' : 'Dismissal Kind'}, axis = 1)
david_dismissal

Dismissal Kind
dismissal_kind                   
bowled                         23
caught                         69
caught and bowled               2
hit wicket                      1
lbw                             4
run out                        10
stumped                         5

**Answer:** David Warner get out by caught mostly.

## Question 5: How many no-balls will be bowled in the match?

#### Solution: Filtering data with SRH and RR Matches and where there is run on no-ball:

In [30]:
wickets_per_match = csk_wickets.groupby('match_id')['player_dismissed'].count()
wickets_per_match = pd.DataFrame(wickets_per_match)
wickets_per_match

player_dismissed
match_id                  
414                      5
435                      3
474                      5
507                      3
521                      4
551                      6
7913                     3
7939                     2
7950                     8
7953                     2
11324                    5
11332                    4

#### Average wickets per match:

In [31]:
avg_wic = wickets_per_match['player_dismissed'].mean()
avg_wic

4.166666666666667

In [10]:
rr_srh_nb = data2.loc[((data2['batting_team'] == "Sunrisers Hyderabad") & (data2['bowling_team'] == "Rajasthan Royals") | (data2['batting_team'] == "Rajasthan Royals") & (data2['bowling_team'] == "Sunrisers Hyderabad")) & (data2['noball_runs'] != 0)]
rr_srh_nb = pd.DataFrame(rr_srh_nb.groupby('match_id')['noball_runs'].count())
rr_srh_nb

noball_runs
match_id             
529                 1
7897                1
11144               1

**Answer:** We see that on an average 1 no ball was bowled between RR and SRH match.